graphql_compiler/schema_transformation/utils.py

```diff
@@ -263,3 +284,84 @@ def check_ast_schema_is_valid(ast):
 
     query_type = get_query_type_name(schema)
     visit(ast, CheckQueryTypeFieldsNameMatchVisitor(query_type))
+
+
+def is_property_field(selection):
+    """Return True if selection is a property field, False if a vertex field or type coercion.
+
+    Args:
+        selection: type Field or InlineFragment. An element occuring inside a SelectionSet
+
+    Returns:
+        True if the selection is a property field, False if it's a vertex field or type coercion
+    """
+    if isinstance(selection, ast_types.InlineFragment):
+        return False
+    if isinstance(selection, ast_types.Field):
+        if (
+            selection.selection_set is None or
+            selection.selection_set.selections is None or
+            selection.selection_set.selections == []
+        ):
+            return True
+        else:
+            return False
+    else:
+        raise AssertionError(
+            u'Input selection "{}" is not of type Field or InlineFragment.'.format(selection)
+        )
+
+
+class CheckQueryIsValidToSplitVisitor(Visitor):
+    """Check the query only has supported directives, and its fields are correctly ordered."""
+    # This is very restrictive for now. Other cases (e.g. tags not crossing boundaries) are
+    # also ok, but temporarily not allowed
+    supported_directives = frozenset(('filter', 'output', 'optional'))
+
+    def enter_Directive(self, node, *args):
+        """Check that the directive is supported."""
+        if node.name.value not in self.supported_directives:
+            raise GraphQLValidationError(
+                u'Directive "{}" is not yet supported, only "{}" are currently '
+                u'supported.'.format(node.name.value, self.supported_directives)
+            )
+
+    def enter_SelectionSet(self, node, *args):
+        """Check property fields occur before vertex fields and type coercions in selection."""
+        past_property_fields = False  # Whether we're seen a vertex field
+        for field in node.selections:
+            if is_property_field(field):
+                if past_property_fields:
+                    raise GraphQLValidationError(
+                        u'In the selections {}, the property field {} occurs after a vertex '
+                        u'field or a type coercion statement, which is not allowed, as all '
+                        u'property fields must appear before all vertex fields.'.format(
+                            node.selections, field
+                        )
+                    )
+            else:
+                past_property_fields = True
```

Comment:
Found this potentially confusing since `past` can either mean we've already seen a property field (which was my original interpretation of this and I was confused because seeing property fields doesn't mean we can't have any more), or it could mean we've gone _past_ any property fields and should expect to see no more. 

How about renaming this to be `seen_vertex_field` or some variant of that?